spec.bs

```diff
@@ -6254,6 +6225,210 @@ an [=origin=] |topLevelOrigin|, and a [=string=] |slotSizeQueryParam|:
+1. Set |signalsUrl|'s [=url/query=] to the result of [=string/concatenating=] |queryParamsList|.
+1. return |signalsUrl|.
+
+</div>
+
+<div algorithm>
+To <dfn>build batched trusted scoring signals url</dfn> given a non-empty [=list=] of [=trusted
+scoring signals requests=] |entriesToBatch|:
+
+1. Let |firstEntry| be |entriesToBatch|[0].
+1. Let |renderURLs| be an empty [=ordered set=] of [=strings=].
+1. Let |adComponentRenderURLs| be an empty [=ordered set=] of [=strings=]
+  |adComponentRenderURLs|
+1. [=list/For each=] |entry| of |entriesToBatch|:
+  1. [=Assert=] that |entry|'s [=trusted scoring signals request/base URL=] is equal to
+    |firstEntry|'s [=trusted scoring signals request/base URL=].
+  1. [=Assert=] that |entry|'s [=trusted scoring signals request/seller experiment group id=] is
+    equal to |firstEntry|'s [=trusted scoring signals request/seller experiment group id=].
+  1. [=Assert=] that |entry|'s [=trusted scoring signals request/top level origin=] is equal to
+    |firstEntry|'s [=trusted scoring signals request/top level origin=].
+  1. [=Assert=] that |entry|'s [=trusted scoring signals request/seller script fetcher=] is equal to
+    |firstEntry|'s [=trusted scoring signals request/seller script fetcher=].
+  1. [=set/Append=] |entry|'s [=trusted scoring signals request/render URL=] to |renderURLs|.
+  1. [=list/Extend=] |adComponentRenderURLs| with |entry|'s [=trusted scoring signals request/
+    ad component URLs=].
+1. Return the result of [=building trusted scoring signals url=] with |firstEntry|'s [=trusted
+  scoring signals request/base URL=], |renderURLs|, |adComponentRenderURLs|, |firstEntry|'s
+  [=trusted scoring signals request/seller experiment group id=],
+  |firstEntry|'s [=trusted scoring signals request/top level origin=].
+
+</div>
+
+<div algorithm>
+To <dfn>check if trusted scoring signals batch honors URL length limit</dfn> given a
+[=trusted scoring signals batcher=] |batcher| and a non-empty [=list=] of [=trusted
+scoring signals requests=] |entriesToBatch|:
+
+1. If the [=list/size=] of |entriesToBatch| is 1, return true.
+1. If |batcher|'s [=trusted scoring signals batcher/url length limit=] is 0, return true.
+1. Return whether the [=string/length=] of the result of [=building batched trusted scoring signals
+  url=] for |entriesToBatch| is &le; |batcher|'s [=trusted scoring signals batcher/url length
+  limit=].
+
+</div>
+
+<div algorithm>
+
+To <dfn>batch and fetch trusted scoring signals</dfn> given a [=trusted scoring signals batcher=]
+|batcher|:
+
+1. Until this object is no longer needed:
+  1. Wait until at least one of the following is true:
+    1. |batcher|'s [=trusted scoring signals batcher/request queue=] [=map/is not empty=].
+    1. |batcher|'s [=trusted scoring signals batcher/request map=] [=map/is not empty=] and some
+      heuristically chosen amount of time has passed.
+  1. Atomically transfer all entries in |batcher|'s [=trusted scoring signals batcher/

```

Comment:
Ah I see. 
1. Add a first step to this algorithm to assert this is running in parallel. (e.g., https://github.com/WICG/turtledove/blob/main/spec.bs#L1635).
2. `Atomically do: 1. [=list/clone=] ... 1. [=list/empty=]...` may be better, since it uses standard methods, instead of "transfer entries", which does not seem clear spec wise.
3. Do you know if there's a good way to spec "atomically"? I vaguelly remembered that you asked about this in one PR. If not, I may want to tag Dominic about their ideas.

Reply:
@domfarolino Is there a way to spec the concept of "atomically"? Not sure if the term itself is enough.

Reply:
I think how this is written is mostly fine, but I have two comments. First, to be extra clear I would add a `Note:` below saying that the effect of this is that if prose elsewhere attempts the modify the [request queue](https://pr-preview.s3.amazonaws.com/morlovich/turtledove/pull/1230.html#trusted-scoring-signals-batcher-request-queue) during this atomic block, the modification will not take place in between the cloning and emptying, meaning we will _always_ empty the same exact items that we clone.... something to that effect would be nice and clear.

Second, I'm curious if this is even needed. The only reason it would be is if (a) the prose doing the reading/cloning/emptying and (b) the prose [doing the _adding_](https://pr-preview.s3.amazonaws.com/morlovich/turtledove/pull/1230.html#ref-for-trusted-scoring-signals-batcher-request-queue) happens in different parallel contexts. I.e., different instances of "in parallel". I think that's the case in the spec right now since we're often just going "in parallel', but if you wanted to serialize these things you could ensure that there is a single parallel queue that both the "adding to the request-queue" task and the "reading from / emptying the request-queue task" get posted to, as to ensure that they both run in parallel from the main thread, but don't run in parallel to _each other_. This is what we do in HTML with the ["session history traversal queue"](https://html.spec.whatwg.org/C#tn-session-history-traversal-queue). It's a named parallel queue so that many things can reference it, and they all run in a sane order, but still in the background. It is somewhat analogous of "the browser process, with regard to session history information".

Reply:
The two execution contexts are indeed distinct and actually parallel.

Reply:
Ah https://html.spec.whatwg.org/multipage/infrastructure.html#starting-a-new-parallel-queue (and the following example) may be also related.

Reply:
> Ah https://html.spec.whatwg.org/multipage/infrastructure.html#starting-a-new-parallel-queue (and the following example) may be also related.

Well only if the two execution contexts are actually the "same", in which they don't need to be distinct in parallel to each other. But they do need to be distinct I suppose, given @morlovich's comment.

Reply:
Right, given Maks's comment (that comment didn't show up yet when I commented 😂 ). So ignore my previous comment then.

Reply:
Note added. Using an actual queue like the parallel queue infra was considered, but it feels awkward to combine with the "just sleep for a bit, too, that's OK" stuff...

Reply:
Can you clarify what you mean by "just sleep for a bit, too, that's OK"? And what bearing it would have on your adoption of a parallel queue? (I'm not saying you should use a parallel queue here I'm just trying to make sure you don't think using one requires arbitrary synchronous sleep steps).

Reply:
So I wasn't really speaking of not using a parallel queue  --- this does after all --- but rather of modelling the fetch request queue it uses on how the "start a new parallel queue" algorithm Qingxin linked to is written --- which is basically continuously spin-dequeueing. On further thought, that approach can actually work here, too, but I think it's less clear than talking about sleeping, since conceptually what this is saying is that implementations waiting for a few milliseconds to merge requests is an acceptable implementation choice. (As is not waiting!)